### ISSUE TITLE
`Sentry\Tracing\Span::__construct()` and `Sentry\Tracing\Span#startChild()` taking around 3% of execution time

### ISSUE THREAD

**Issue by Ocramius** [getsentry/sentry-php#1273]
### Environment

 * relatively boring `symfony/*:^6` application with `doctrine/dbal` and `doctrine/orm`
 * `sentry/sentry-symfony:4.2.5`
 * `sentry/sdk:3.1.1` (yes, I know newer versions are out - checked diffs, didn't find anything related to this issue)

### Steps to Reproduce

Ran a full integration test suite with `XDEBUG_MODE=profile vendor/bin/phpunit`, produced a huge `cachegrind.out` file that I analyzed ( https://twitter.com/Ocramius/status/1479844147427172360 _

### Expected Result

Sentry should pretty much be non-existing in such a profile :D 

### Actual Result

 * `Sentry\Tracing\Span::__construct()` is one of the most time-consuming operations
 * `Sentry\Tracing\Span#startChild()` is one of the most time-consuming operations
 
 Here's a screenshot for clarity:

![Selection_877](https://user-images.githubusercontent.com/154256/148654480-9e1221c7-be2b-4581-afa3-6db49e5a8e73.png)

In practice, this means that `Sentry\Tracing\Span` is potentially heavier than most ORM operations, which I find quite weird.
Obviously, it is called 850000 times, in the context of the test suite, which matches query start/end, plus HTTP request start/end (where I'd expect the spans to be placed, kinda).

Looking at the implementation itself, I don't see anything evidently "high consumption", other than perhaps querying the system for `microtime(true)`:

https://github.com/getsentry/sentry-php/blob/c186c44c32899ad0cf5b4e942d71035f01b87b64/src/Tracing/Span.php#L81-L111
https://github.com/getsentry/sentry-php/blob/c186c44c32899ad0cf5b4e942d71035f01b87b64/src/Tracing/Span.php#L388-L409

I wonder if:

 * accessing `$context` could be slow?
 * cloning `$context` could be slow?
 * `microtime()` calls are not cached (because of missing FQN reference)?

Obviously, the resolution of the profile I have is low, because it covers a test suite run with thousands of tests, but I figured it would be a good idea to report this here, since I really wasn't expecting the `Span` API to be part of the higher-up time consumers.

**Comment by ste93cry** [comment]
There could be some lines that could be optimized, for example to avoid calling the getters twice in the `if` statements of `Span::__construct()`, but it seems weird to me that it's the problem as there is for sure a lot of other code much heavier than this. Maybe, out of curiosity, could you try to profile after removing the call to `microtime(true)` and see if there is any difference? The other thing that I see is that we're calling `(Trace|Span)::generate()` which generates an UUID, and if you don't have installed `ext-uuid` the fallback is the Symfony polyfill. Maybe is it that one that slow down everything?

**Comment by Ocramius** [comment]
Nope, UUID extension installed, and polyfill categorically excluded.

Note that this is "self" time, not inclusive time.

**Comment by Jean85** [comment]
IMHO we could optimize out all 3 of the: Trace ID generation SpanId generation and `microtime`. As I can see, all 3 of them can be overwritten by context.

Another possible optimization could be extracting all the values from the context into a read-only value object, so that each span could copy only the reference to it, and if it doesn't change much, we don't copy data nor create more object instances.

**Comment by ste93cry** [comment]
To be honest it's pretty strange to me that some `if` statements are the cause of such an high execution time. But, as I suggested before, maybe @Ocramius could you try profiling without the call to `microtime(true)` and see if it makes any difference?

**Comment by Ocramius** [comment]
I'll certainly try producing a much more reduced profile, so I get a better resolution around it :)

**Comment by Jean85** [comment]
> To be honest it's pretty strange to me that some `if` statements are the cause of such an high execution time. But, as I suggested before, maybe @Ocramius could you try profiling without the call to `microtime(true)` and see if it makes any difference?

IMO it's not the `if` but the (unuseful) regeneration of IDs and microtime. If it's not the first span and we have a context, the `if`s pass, and the data is overridden. If we use an `else` we can reverse that logic.

@Ocramius maybe we can try to use php-bench on `Span::__construct`?

**Comment by Jean85** [comment]
I've tried PHPBench in #1274

**Comment by Ocramius** [comment]
A PHPBench approach is indeed much more valuable: good call!

**Comment by Jean85** [comment]
Bench and optimization merged. Let us know if that's enough, or we can reopen this otherwise.

**Comment by Ocramius** [comment]
Let's close here for now: it takes many hours to generate one of those memory profiles, so I'll see  when I can get one 👍

**Comment by Ocramius** [comment]
Thanks @Jean85! Stellar work!
